src/wp-includes/media.php

```diff
@@ -949,27 +961,22 @@ function wp_get_registered_image_subsizes() {
  *     @type bool   $3 Whether the image is a resized image.
  * }
  */
-function wp_get_attachment_image_src( $attachment_id, $size = 'thumbnail', $icon = false ) {
-	// Get a thumbnail or intermediate image if there is one.
-	$image = image_downsize( $attachment_id, $size );
-	if ( ! $image ) {
-		$src = false;
-
-		if ( $icon ) {
-			$src = wp_mime_type_icon( $attachment_id );
+function wp_get_attachment_image_src( $attachment_id, $size = 'thumbnail', $args = array() ) {
+	$mime_type = false;
+	$icon      = false;
 
-			if ( $src ) {
-				/** This filter is documented in wp-includes/post.php */
-				$icon_dir = apply_filters( 'icon_dir', ABSPATH . WPINC . '/images/media' );
+	if ( is_array( $args ) && isset( $args['mime_type'] ) ) {
+		$mime_type = $args['mime_type'];
+	}
 
-				$src_file               = $icon_dir . '/' . wp_basename( $src );
-				list( $width, $height ) = wp_getimagesize( $src_file );
-			}
-		}
+	if ( is_array( $args ) && isset( $args['icon'] ) || is_bool( $args ) ) {
+		$icon = isset( $args['icon'] ) ? $args['icon'] : $args;
+	}
```

Comment:
This is overly complicated as there is a duplicate `isset` check. Also, we should be more forgiving for the backward compatibility fix, any non-array `$args` can reliably be interpreted as `$icon`.
```suggestion
	if ( is_array( $args ) && isset( $args['icon'] ) ) {
		$icon = $args['icon'];
	} else if ( ! is_array( $args ) ) {
		$icon = (bool) $args;
	}
```